Support disk space reclaim estimation

dotnwat opened the pull request:
The primary changes in this pull request are:

* Parallel removal of persistent segment files
* Reporting of data removed from disk on segment removal
* Estimation of data available for reclaim by GC under 2 scenarios:
  * Normal retention policy
  * Safe, but violating retention policy (e.g. uploaded to cloud)

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

dotnwat commented:
Failures
* https://github.com/redpanda-data/redpanda/issues/7756
* https://github.com/redpanda-data/redpanda/issues/8496

emaxerrno commented:
THIS IS COOOOOL

VladLazar commented on src/v/storage/disk_log_impl.h:
```diff
@@ -34,6 +34,22 @@
+ *
+ * available: number of bytes that could be safely reclaimed if the
+ *            retention policy was ignored. for example reclaiming
+ *            past the retention limits if the data being relcaimed
```
nit: relcaimed -> reclaimed

VladLazar commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1697,4 +1697,74 @@ log make_disk_backed_log(
+    for (auto& seg : _segs) {
+        if (seg->offsets().dirty_offset < retention_offset) {
+            retention_segments.push_back(seg);
+        } else if (seg->offsets().dirty_offset < max_collectible) {
```
Should the condition here be: `seg->offsets().dirty_offset <= max_collectible`? Same for the if above.

dotnwat replied:
@VladLazar i don't think so, but i'm interested if you think otherwise. in `remove_prefix_full_segments` the predicate in `do_until(p, remove_front_segment)` is `p = _segs.front()->offsets().dirty_offset >= cfg.start_offset`, so the negation of that would be strictly `<`. wdyt?

VladLazar replied:
`log_eviction_stm` determines the offset at which the snapshot will occur [here](https://github.com/redpanda-data/redpanda/blob/8a9d43faa4b0d998eb6cd79521090b1733eca0e1/src/v/raft/log_eviction_stm.cc#L73-L80). Let's assume that's offset `X`. Then, raft requests truncates the log at `X + 1` [here](https://github.com/redpanda-data/redpanda/blob/8a9d43faa4b0d998eb6cd79521090b1733eca0e1/src/v/raft/consensus.cc#L2217-L2218). Finally, all segments with `dirty_offset < X + 1` are removed [here](https://github.com/redpanda-data/redpanda/blob/8a9d43faa4b0d998eb6cd79521090b1733eca0e1/src/v/storage/disk_log_impl.cc#L1265-L1277).

`retention_offset` in your code is offset `X` (what `log_eviction_stm` comes up with, so the comparison on line 1742 should use `<=`  (i think).

dotnwat replied:
You're right. I missed `model::next_offset(last_included_index)` when I was tracing the code yesterday. Thanks!

VladLazar commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1697,4 +1697,74 @@ log make_disk_backed_log(
+        co_return reclaim_size_limits();
+    }
+
+    /*
+     * evicting data based on the retention policy is a coordinated effort
+     * between disk_log_impl housekeeping and the raft/eviction_stm. it works
+     * roughly as follows:
+     *
+     * 1. log computes ideal target offset for reclaim: max_offset computed by
+     *    the call to retention_offset(cfg).
+     *
+     * 2. log notifies the eviction stm which computes: min(max_offset, clamp)
+     *    where clamp is any additional restrictions, such as what has been
+     *    uploaded to cloud storage.
+     *
+     * 3. raft attempts to prefix truncate the log at the clamped offset after
+     *    it has prepared and taken any necessary snapshots.
+     *
+     * Because this effort is split between the log, raft, and eviction_stm we
+     * end up having to duplicate some of the logic here, such as handling the
+     * max collectible offset from stm. A future refactoring may consider moving
+     * more of the retention controls into a higher level location.
+     */
```
💘

VladLazar commented on src/v/storage/tests/utils/disk_log_builder.cc:
```diff
@@ -137,6 +137,19 @@ ss::future<> disk_log_builder::gc(
+ss::future<reclaim_size_limits> disk_log_builder::gc_estimate(
+  model::timestamp collection_upper_bound,
+  std::optional<size_t> max_partition_retention_size) {
+    ss::abort_source as;
```
This `abort_source` is not used.

VladLazar commented on src/v/storage/tests/log_retention_tests.cc:
```diff
@@ -124,6 +124,10 @@ FIXTURE_TEST(retention_test_size_time, gc_fixture) {
         offset += model::offset(num_records);
     }
 
+    // there is only an active segment. nothing should be reclaimable
```
You asked me about this out of band, but we don't preserve the invariant that one segment is always on disk.

dotnwat replied:
cool, thanks. didn't know that.

VladLazar reviewed (commented):
neis

dotnwat commented:
Force push https://github.com/redpanda-data/redpanda/compare/dce97abb35c3c4f925884acd7bb7075dfddc1adf..41d2d464038b682c4410f81365be538e57f5dd1b

* Fixed conditional
* Fixed spelling error
* Removed unused abort source

VladLazar reviewed (approved):
LGTM. I'm curious if you plan to expose this estimation outside of Redpanda (via the admin api maybe). I think it would be useful.

dotnwat commented:
> LGTM. I'm curious if you plan to expose this estimation outside of Redpanda (via the admin api maybe). I think it would be useful.

Probably. The next set of PRs will further enhance this information so we can start making additional decisions. Seems natural to want to expose it eventually if even initially to drive testing.
